### DISCUSSION TITLE
Support different SSO Auth Methods like OAuth/LDAP/OIDC with an auth proxy

### DISCUSSION THREAD

**Discussion by Vad1mo** [plausible/analytics#383] [Anything Else]
In order to allow users to login to plausible with different types of SSO there are two option to accomplish this.

1. Implement the different SSO methods in your application
2. Add Support for third party auth proxies. **(<- This feature request)**

## Feature request
Add  auth proxies support to plausible allowing to support a plethora of authentication options without polluting the code with Auth logic and code maintenance for the team. 

**How would you like it to work**

The implementation is quite simple:

1. Plausible needs an option to trust auth header containing the request header with the user id or email eg.: `X-Auth-UserId` 
2. In order to distinguish if it's a user or admin a second group header can be evaluated eg.: `X-Auth-Groups` runa test against a regex. Alternatively a hard coded setting in the config could also determine who is the admin.

![image](https://user-images.githubusercontent.com/1492007/97163946-b46a0880-1781-11eb-8210-c3ef5b652f6c.png)

This was plausible could easily support of sorts of auth mechanisms without implementing each one of it.
There is a whole ecosystem that would be instantly available cloud providers and web servers  like AWS, GCP, Azure, Auth0, Nginx, Apache all support this type of auth method.

here some examples, how others are doing it. 
- https://grafana.com/docs/grafana/latest/auth/auth-proxy/

**Update:** SSO is now available! Details at https://plausible.io/docs/sso

**Comment by alexanderadam** [comment]
Just in case somebody wants to give it a go:

Sign ins, for the well known I18n Elixir app [Accent, can be setup for external providers via few environment variables](https://github.com/mirego/accent#authentication-setup) and since they're using [Überauth](https://github.com/ueberauth/ueberauth#%C3%BCberauth) adding [another provider strategy like GitLab](https://github.com/mirego/accent/pull/264/files#diff-e43c5cbf91db7a8062b6cb860cbf118296c1b4c7ee32fdcf702e54234ba38092) seem to be straightforward.

[Überauth seem to support all popular variants from self hosted and external providers](https://github.com/ueberauth/ueberauth/wiki/List-of-Strategies).

And with external authentication (_i.e. with OAuth/OpenID Connect_), people could also activating 2FA on the authentication app, hence this would clearly solve [issue 156](https://github.com/plausible/analytics/discussions/156) for some users.

**Comment by bkosborne** [comment]
Would this proposed implementation work with SAML providers, like Shiboleth?

**Comment by lorenz** [comment]
#3170 is a working OIDC implementation, but still a bit rough around the edges. Still need to add automated user creation/profile metadata updates and some additional security things.

**Comment by metmarkosaric** [comment]
**Update:** SSO is now available! Details at https://plausible.io/docs/sso

**Reply by sebdanielsson** [reply]
No OIDC? :/

**Reply by metmarkosaric** [reply]
@sebdanielsson unfortunately not. we only support SAML 2.0 currently. there's a dedicated thread for OIDC on our feedback board now: https://feedback.plausible.io/643

**Comment by stepbrobd** [comment]
Just updated my self-hosted instance to 3.0.1, and SSO don’t seem to be available for team account

Does it mean that SSO/OIDC (if implemented) will not be available for self-hosted instances? https://github.com/plausible/analytics/pull/3170#issuecomment-2143976758

It seems that the current SSO feature implementation is guarded by under a flag (https://github.com/plausible/analytics/commit/bca78169bb026755b29a3e0bc3723feae632af68), are self hosters allowed to set it to true in their deployments?
